src/storage.rs

```diff
@@ -1,85 +1,112 @@
 use serde::de::DeserializeOwned;
 use serde::Serialize;
 use serde_json::{from_str, to_string};
+use std::fmt;
 use std::fs::{File, OpenOptions};
-use std::io::{Read, Write, Seek, SeekFrom};
+use std::io::{Read, Seek, SeekFrom, Write};
 use std::ops::{Deref, DerefMut};
-use std::fmt;
 
+// General traits
+pub trait StorageReadGuard<'a, T>: Deref<Target = T> {}
+pub trait StorageWriteGuard<'a, T>: DerefMut<Target = T> {}
+
+pub trait Storage<'a>: fmt::Display {
+    type S;
+
+    fn read(&'a self) -> Box<StorageReadGuard<'a, Self::S, Target = Self::S> + 'a>;
+    fn write(&'a mut self) -> Box<StorageWriteGuard<'a, Self::S, Target = Self::S> + 'a>;
+}
+
+// Disk Storage-related structs and impls
 /// TODO: Write some docs
 #[derive(Debug)]
-pub struct StorageReadGuard<'a, T: Serialize + DeserializeOwned + 'a> {
-    cache: &'a Storage<T>,
+pub struct DiskStorageReadGuard<'a, T: Serialize + DeserializeOwned + 'a> {
+    __storage: &'a DiskStorage<T>,
```

Comment:
Why are you using two underscores in front of the variable name here? The standing convention is to let the compiler enforce public vs. private.

Reply:
Borrowed from https://doc.rust-lang.org/src/std/sync/rwlock.rs.html#121, but I can't seem to find a solid reason that they do it there, so I'll just switch back to no leading underscores